firebase-firestore/src/androidTest/java/com/google/firebase/firestore/QueryTest.java

```diff
@@ -80,6 +81,283 @@ public void testLimitQueriesUsingDescendingSortOrder() {
+
+    waitFor(collection.add(map("k", "e", "sort", -1)));
+    waitFor(testCounter);
+
+    assertEquals(2, snapshots.size());
+    data = querySnapshotToValues(snapshots.get(1));
+    assertEquals(asList(map("k", "a", "sort", 0L), map("k", "e", "sort", -1L)), data);
+  }
+
+  // Two queries that mapped to the same target ID are referred to as
+  // "mirror queries". An example for a mirror query is a limitToLast()
+  // query and a limit() query that share the same backend Target ID.
+  // Since limitToLast() queries are sent to the backend with a modified
+  // orderBy() clause, they can map to the same target representation as
+  // limit() query, even if both queries appear separate to the user.
+  @Test
+  public void testListenToMirrorQueries() {
+    CollectionReference collection =
+        testCollectionWithDocs(
+            map(
+                "a", map("k", "a", "sort", 0),
+                "b", map("k", "b", "sort", 1),
+                "c", map("k", "c", "sort", 1),
+                "d", map("k", "d", "sort", 2)));
+
+    List<QuerySnapshot> limitSnapshots = new ArrayList<>();
+    Semaphore limitTestCounter = new Semaphore(0);
+    Query limit = collection.limit(2).orderBy("sort", Direction.ASCENDING);
+
+    limit.addSnapshotListener(
+        (snapshot, error) -> {
+          assertNull(error);
+          limitSnapshots.add(snapshot);
+          limitTestCounter.release();
+        });
+
+    List<QuerySnapshot> limitToLastSnapshots = new ArrayList<>();
+    Semaphore limitToLastTestCounter = new Semaphore(0);
+    Query limitToLast = collection.limitToLast(2).orderBy("sort", Direction.DESCENDING);
+
+    limitToLast.addSnapshotListener(
+        (snapshot, error) -> {
+          assertNull(error);
+          limitToLastSnapshots.add(snapshot);
+          limitToLastTestCounter.release();
+        });
+
+    waitFor(limitTestCounter);
+    assertEquals(1, limitSnapshots.size());
+    List<Map<String, Object>> data = querySnapshotToValues(limitSnapshots.get(0));
+    assertEquals(asList(map("k", "a", "sort", 0L), map("k", "b", "sort", 1L)), data);
+
+    waitFor(limitToLastTestCounter);
+    assertEquals(1, limitToLastSnapshots.size());
+    data = querySnapshotToValues(limitToLastSnapshots.get(0));
+    assertEquals(asList(map("k", "b", "sort", 1L), map("k", "a", "sort", 0L)), data);
+
+    waitFor(collection.add(map("k", "e", "sort", -1)));
+
+    waitFor(limitTestCounter);
+    assertEquals(2, limitSnapshots.size());
+    data = querySnapshotToValues(limitSnapshots.get(1));
+    assertEquals(asList(map("k", "e", "sort", -1L), map("k", "a", "sort", 0L)), data);
+
+    waitFor(limitToLastTestCounter);
+    assertEquals(2, limitToLastSnapshots.size());
+    data = querySnapshotToValues(limitToLastSnapshots.get(1));
+    assertEquals(asList(map("k", "a", "sort", 0L), map("k", "e", "sort", -1L)), data);
+  }
+
+  @Test
+  public void testUnlistenFromMirrorQueries() {
```

Comment:
The only reason is if one of them failed, it might be easier to figure out where it failed since some of them cover smaller scope. 

That said, I don't think this justifies the amount of duplication here..I merged them into one.